src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php

```diff
@@ -48,21 +49,27 @@ class UserBadge implements BadgeInterface
      * is thrown). If this is not set, the default user provider will be used with
      * $userIdentifier as username.
      */
-    public function __construct(string $userIdentifier, callable $userLoader = null)
+    public function __construct(string $userIdentifier, callable $userLoader = null, array $attributes = null)
     {
         if (\strlen($userIdentifier) > self::MAX_USERNAME_LENGTH) {
             throw new BadCredentialsException('Username too long.');
         }
 
         $this->userIdentifier = $userIdentifier;
         $this->userLoader = $userLoader;
+        $this->attributes = $attributes;
     }
 
     public function getUserIdentifier(): string
     {
         return $this->userIdentifier;
     }
 
+    public function getAttributes(): ?array
```

Comment:
What's the purpose of `null`? Could the constructor arg default to an empty list instead? That would allow removing the null checks so less boilerplate

Reply:
It prevents calling the user loader with a useless extra argument:
```php
public function getUser(): UserInterface
{
    // ...
    if (null === $this->getAttributes()) {
        $user = ($this->userLoader)($this->userIdentifier);
    } else {
        $user = ($this->userLoader)($this->userIdentifier, $this->getAttributes());
    }
    // ...
}
```